### PR TITLE
Validate pods in Submariner Deployment

### LINKED ISSUE #1141: Validate  all the Submariner Pods
**What would you like to be added**:
Verify that all the Submariner Pods (expected configuration can be read from the Submariner CR) are in Running state.

**Why is this needed**:
This would help in troubleshooting submariner issues.

### PR DESCRIPTION
**Pull request by aswinsuryan**
Fixes: #1141

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

### CODE REVIEW

File: pkg/subctl/cmd/validateSubmariner.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	subMObj, err := submarinerResourceClient.Get("submariner", metav1.GetOptions{})
+	if err != nil {
+		exitOnError("failed to validate submariner cr due to", err)
+	}
+	submariner := &v1alpha1.Submariner{}
+	submarinerScheme := subMScheme.Scheme
+	err = submarinerScheme.Convert(subMObj, submariner, nil)
+	if err != nil {
+		exitOnError("failed to validate submariner cr due to", err)
+	}
```
Comment (tpantelis) [inline pkg/subctl/cmd/validateSubmariner.go]: Why not use the Submariner client set instead of dynamic here and below?

**Review by tpantelis — CHANGES_REQUESTED** [review]
Seems like this should be built on top of https://github.com/submariner-io/submariner-operator/pull/1139.

File: pkg/subctl/cmd/validate.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		Short: "Validate Submariner deployment and report any issues",
+		Long: "This command validates a Submariner deployment and reports any issues if \n" +
+			"1. Submariner pre-requisites are missing on the cluster\n" +
+			"2. Submariner is deployed on an unsupported k8s cluster\n" +
+			"3. Submariner Pods are in an error state\n",
+		Run: validateSubmariner,
+	}
+	validateK8sCmd = &cobra.Command{
+		Use:   "k8s",
+		Short: "Validate if K8s Cluster configuration is supported by Submariner",
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validate.go]: This PR depends on the other PR - https://github.com/submariner-io/submariner-operator/pull/1139 for validate.go file. As part of the review comments, I modified the file validate.go in my PR. We can avoid reviewing it again in this PR and use the one from my PR during merging.

File: pkg/subctl/cmd/validateSubmariner.go (hunk 1 above)
Comment (sridhargaddam) [reply pkg/subctl/cmd/validateSubmariner.go]: Submariner clientset would be simpler to use.

File: pkg/subctl/cmd/validateSubmariner.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+		fmt.Printf("\nValidating Service Discovery Pods for cluster: %q", clusterName)
+		// Check lighthouse-agent
+		CheckDeployment(localClient, operatorNamespace, "submariner-lighthouse-agent")
+
+		// Check ligthouse-coreDNS
+		CheckDeployment(localClient, operatorNamespace, "submariner-lighthouse-coredns")
+	}
+	// Check if globalnet components are deployed and running if enabled
+	if submariner.Spec.GlobalCIDR != "" {
+		fmt.Printf("\nValidating Global Pods for cluster: %q", clusterName)
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validateSubmariner.go]: Globalnet

Comment (aswinsuryan) [reply pkg/subctl/cmd/validateSubmariner.go]: Done

File: pkg/subctl/cmd/validateSubmariner.go (hunk 1 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validateSubmariner.go]: Done.

File: pkg/subctl/cmd/validateSubmariner.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		CheckDaemonset(kubeClientSet, operatorNamespace, "submariner-globalnet")
+	}
+	fmt.Println()
+}
+
+func CheckDeployment(k8sClient kubernetes.Interface, namespace, deploymentName string) {
+	deployment, err := k8sClient.AppsV1().Deployments(namespace).Get(deploymentName, metav1.GetOptions{})
+	if err != nil {
+		err = fmt.Errorf("validation %q of deployment failed due to error: %v", deploymentName, err)
+		exitOnError("failed", err)
```
Comment (maayanf24) [inline pkg/subctl/cmd/validateSubmariner.go]: when ever there is a call to `exitOnError`, the command will stop and not continue to validate other components, or other clusters

**Review by maayanf24 — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/validate.go (hunk 2 above)
Comment (sridhargaddam) [reply pkg/subctl/cmd/validate.go]: The base PR is now merged. Please rebase to get the latest changes. We modified this code to be aligned with `subctl show`

File: pkg/subctl/cmd/validateSubmariner.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+	// Check if submariner components are deployed and running
+	fmt.Printf("\nValidating Submariner Pods for cluster: %q", clusterName)
+	CheckDaemonset(kubeClientSet, operatorNamespace, "submariner-gateway")
+
+	CheckDaemonset(kubeClientSet, operatorNamespace, "submariner-routeagent")
+
+	// Check if service-discovery components are deployed and running if enabled
+	if submariner.Spec.ServiceDiscoveryEnabled {
+		fmt.Printf("\nValidating Service Discovery Pods for cluster: %q", clusterName)
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validateSubmariner.go]: Please use status to indicate success or failures

```
status.QueueSuccessMessage(message)
status.End(cli.Success)
```

**Review by sridhargaddam — CHANGES_REQUESTED** [review]
Thanks for the PR @aswinsuryan

File: pkg/subctl/cmd/validate.go (hunk 2 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate.go]: Done.

File: pkg/subctl/cmd/validateSubmariner.go (hunk 4 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validateSubmariner.go]: changed to status package methods.

File: pkg/subctl/cmd/validateSubmariner.go (hunk 5 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validateSubmariner.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+		// Check ligthouse-coreDNS
+		CheckDeployment(kubeClientSet, operatorNamespace, "submariner-lighthouse-coredns")
+	}
+	// Check if globalnet components are deployed and running if enabled
+	if submariner.Spec.GlobalCIDR != "" {
+		CheckDaemonset(kubeClientSet, operatorNamespace, "submariner-globalnet")
+	}
+	message := "All Submariner pods are up and running"
+	status.QueueFailureMessage(message)
```
Comment (maayanf24) [inline pkg/subctl/cmd/validate_deployment.go]: in this case the status should be Success message

**Review by maayanf24 — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/validate_deployment.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+	CheckDaemonset(kubeClientSet, operatorNamespace, "submariner-gateway")
+
+	CheckDaemonset(kubeClientSet, operatorNamespace, "submariner-routeagent")
+
+	// Check if service-discovery components are deployed and running if enabled
+	if submariner.Spec.ServiceDiscoveryEnabled {
+		// Check lighthouse-agent
+		CheckDeployment(kubeClientSet, operatorNamespace, "submariner-lighthouse-agent")
+
+		// Check ligthouse-coreDNS
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
		// Check lighthouse-coreDNS
```

File: pkg/subctl/cmd/validate_deployment.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	message := "All Submariner pods are up and running"
+	status.QueueFailureMessage(message)
+}
+
+func CheckDeployment(k8sClient kubernetes.Interface, namespace, deploymentName string) {
+	deployment, err := k8sClient.AppsV1().Deployments(namespace).Get(deploymentName, metav1.GetOptions{})
+	if err != nil {
+		message := fmt.Sprintf("validation %q of deployment failed due to error: %v",
+			deploymentName, err)
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
		message := fmt.Sprintf("validation of deployment %q failed due to error: %v",
			deploymentName, err)
```

File: pkg/subctl/cmd/validate_deployment.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+		return
+	}
+
+	if deployment.Status.AvailableReplicas != *deployment.Spec.Replicas {
+		message := fmt.Sprintf("the configured number of replicas are not running for %q", deploymentName)
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validate_deployment.go]: IMO its better to use first letter as Capital letter for the messages.
```suggestion
		message := fmt.Sprintf("The configured number of replicas are not running for %q", deploymentName)
```

File: pkg/subctl/cmd/validate_deployment.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+		return
+	}
+}
+
+func CheckDaemonset(k8sClient kubernetes.Interface, namespace, daemonSetName string) {
+	daemonSet, err := k8sClient.AppsV1().DaemonSets(namespace).Get(daemonSetName, metav1.GetOptions{})
+	if err != nil {
+		message := fmt.Sprintf("validation %q of daemonset failed due to error: %v", daemonSetName, err)
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validate_deployment.go]: Same as above

File: pkg/subctl/cmd/validate_deployment.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+		return
+	}
+
+	if daemonSet.Status.CurrentNumberScheduled != daemonSet.Status.DesiredNumberScheduled {
+		message := fmt.Sprintf("the desried number of daemonsets are not running for %q", daemonSetName)
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
		message := fmt.Sprintf("The desired number of daemonsets are not running for %q", daemonSetName)
```

File: pkg/subctl/cmd/validate_deployment.go (hunk 6 above)
Comment (sridhargaddam) [reply pkg/subctl/cmd/validate_deployment.go]: I think you should check if there was any error while validating the Deployments/DaemonSets and report the status accordingly.

File: pkg/subctl/cmd/validate_deployment.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/submariner-io/submariner-operator/pkg/internal/cli"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var validatePodsCmd = &cobra.Command{
+	Use:   "pods",
+	Short: "Validate the submariner pods",
+	Long:  "This command validates all the submariner pods are running",
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
	Long:  "This command checks that all the submariner pods are running",
```

File: pkg/subctl/cmd/validate_deployment.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+func init() {
+	validateCmd.AddCommand(validatePodsCmd)
+}
+
+func validatePods(cmd *cobra.Command, args []string) {
+	configs, err := getMultipleRestConfigs(kubeConfig, kubeContext)
+	exitOnError("error getting REST config for cluster", err)
+
+	for _, item := range configs {
+		message := fmt.Sprintf("Validating Submariner Pods in %q", item.clusterName)
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]: We don't capitalize these in the help messages - either way we should be consistent.

```suggestion
		message := fmt.Sprintf("Validating submariner pods in %q", item.clusterName)
```

File: pkg/subctl/cmd/validate_deployment.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+	submariner, err := submarinerResourceClient.SubmarinerV1alpha1().Submariners(operatorNamespace).
+		Get("submariner", metav1.GetOptions{})
+
+	if err != nil {
+		message := fmt.Sprintf("failed to validate submariner cr due to: %v", err)
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+		return
+	}
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
        submariner := getSubmarinerResource(config)

        if submariner == nil {
	    status.QueueWarningMessage(submMissingMessage)
	    status.End(cli.Success)
	    return
        }
```

File: pkg/subctl/cmd/validate_deployment.go [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+	kubeClientSet, err := kubernetes.NewForConfig(config)
+
+	if err != nil {
+		message := fmt.Sprintf("error creating kubernetes clientset: %v", err)
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+		return
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
                exitOnError("error creating Kubernetes client", err)
```

File: pkg/subctl/cmd/validate_deployment.go (hunk 8 above)
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
		message := fmt.Sprintf("Error obtaining Deployment %q: %v", deploymentName, err)
```

File: pkg/subctl/cmd/validate_deployment.go (hunk 9 above)
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]: Also print the number available/desired:

```suggestion
		message := fmt.Sprintf("The desired number of replicas for Deployment %q (%d) does not match the actual number running (%d)", deploymentName, replicas, deployment.Status.AvailableReplicas)
```

File: pkg/subctl/cmd/validate_deployment.go [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+	deployment, err := k8sClient.AppsV1().Deployments(namespace).Get(deploymentName, metav1.GetOptions{})
+	if err != nil {
+		message := fmt.Sprintf("validation %q of deployment failed due to error: %v",
+			deploymentName, err)
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+		return
+	}
+
+	if deployment.Status.AvailableReplicas != *deployment.Spec.Replicas {
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]: `deployment.Spec.Replicas` could b e nil in which case it defaults to 1.

```suggestion
        replicas := 1
        if deployment.Spec.Replicas != nil {
            replicas = *deployment.Spec.Replicas
        }
        
	if deployment.Status.AvailableReplicas != replicas {
```

File: pkg/subctl/cmd/validate_deployment.go (hunk 6 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Added a check and will report success only if all tests passed

File: pkg/subctl/cmd/validate_deployment.go (hunk 11 above)
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
		message := fmt.Sprintf("The desired number of nodes that are  running pods for DaemonSet %q (%d) does not match the actual number (%d)", daemonSetName)
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/validate_deployment.go (hunk 11 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go (hunk 16 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go (hunk 9 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go (hunk 8 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go (hunk 15 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go (hunk 14 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go (hunk 13 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go (hunk 12 above)
Comment (aswinsuryan) [reply pkg/subctl/cmd/validate_deployment.go]: Done.

File: pkg/subctl/cmd/validate_deployment.go [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+	daemonSet, err := k8sClient.AppsV1().DaemonSets(namespace).Get(daemonSetName, metav1.GetOptions{})
+	if err != nil {
+		message := fmt.Sprintf("Error obtaining Daemonset %q: %v", daemonSetName, err)
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+		return false
+	}
+
+	if daemonSet.Status.CurrentNumberScheduled != daemonSet.Status.DesiredNumberScheduled {
+		message := fmt.Sprintf("The desired number of nodes that are  running pods for DaemonSet %q (%d) " +
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
		message := fmt.Sprintf("The desired number of nodes that are running pods for DaemonSet %q (%d) " +
```

**Review by tpantelis — CHANGES_REQUESTED** [review]
one more minor thing

**Review by tpantelis — APPROVED** [review]

File: pkg/subctl/cmd/validate_deployment.go [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+	daemonSet, err := k8sClient.AppsV1().DaemonSets(namespace).Get(daemonSetName, metav1.GetOptions{})
+	if err != nil {
+		message := fmt.Sprintf("Error obtaining Daemonset %q: %v", daemonSetName, err)
+		status.QueueFailureMessage(message)
+		status.End(cli.Failure)
+		return false
+	}
+
+	if daemonSet.Status.CurrentNumberScheduled != daemonSet.Status.DesiredNumberScheduled {
+		message := fmt.Sprintf("The desired number of nodes that are running pods for DaemonSet %q (%d)"+
```
Comment (sridhargaddam) [inline pkg/subctl/cmd/validate_deployment.go]:
```suggestion
		message := fmt.Sprintf("The desired number of running pods for DaemonSet %q (%d)"+
```

For some of our DaemonSets like gateway, globalnet etc we use Selectors. Hence they do not run on each and every node of the Cluster. IMO, its better to remote reference to nodes in the message and just say that desired number of running pods does not match with actual number.

**Review by sridhargaddam — APPROVED** [review]
Thanks for addressing the review comments @aswinsuryan

**Review by maayanf24 — APPROVED** [review]

**Pull request merged**